pyiceberg/table/__init__.py

```diff
@@ -791,6 +792,76 @@ class CommitTableResponse(IcebergBaseModel):
     metadata_location: str = Field(alias="metadata-location")
 
 
+class TableCommitRetry:
+    """Decorator for building the table commit retry controller."""
+
+    num_retries = "commit.retry.num-retries"
+    num_retries_default: int = 4
+    min_wait_ms = "commit.retry.min-wait-ms"
+    min_wait_ms_default: int = 100
+    max_wait_ms = "commit.retry.max-wait-ms"
+    max_wait_ms_default: int = 60000  # 1 min
+    total_timeout_ms = "commit.retry.total-timeout-ms"
+    total_timeout_ms_default: int = 1800000  # 30 mins
+
+    def __init__(self, func: Callable[..., Any], properties_attribute: str = "properties") -> None:
+        self.properties_attr: str = properties_attribute
+        self.func: Callable[..., Any] = func
+        self.loaded_properties: Properties = {}
+
+    def __get__(self, instance: Any, owner: Any) -> Callable[..., Any]:
+        """Return the __call__ method with the instance caller."""
+        return partial(self.__call__, instance)
+
+    def __call__(self, instance: Any, *args: Any, **kwargs: Any) -> Any:
+        """Run function with the retrying controller on the caller instance."""
+        self.loaded_properties = getattr(instance, self.properties_attr)
+        try:
+            for attempt in self.build_retry_controller():
+                with attempt:
+                    result = self.func(instance, *args, **kwargs)
+        except RetryError as err:
+            raise Exception from err.reraise()
+        else:
+            return result
+
+    @property
+    def table_properties(self) -> Properties:
+        """Get the table properties from the instance that is calling this decorator."""
+        return self.loaded_properties
+
+    def build_retry_controller(self) -> Retrying:
+        """Build the retry controller."""
+        return Retrying(
+            stop=(
+                stop_after_attempt(self.get_config(self.num_retries, self.num_retries_default))
+                | stop_after_delay(
+                    datetime.timedelta(milliseconds=self.get_config(self.total_timeout_ms, self.total_timeout_ms_default))
+                )
+            ),
+            wait=wait_exponential(min=self.get_config(self.min_wait_ms, self.min_wait_ms_default) / 1000.0),
+            retry=retry_if_exception_type(CommitFailedException),
+        )
+
+    def get_config(self, config: str, default: int) -> int:
+        """Get config out of the properties."""
+        return self.to_int(self.table_properties.get(config, ""), default)
```

Comment:
If the key doesn't exists we try to convert the empty string. How about throwing in some walrus `:=`:

```suggestion
        return self.to_int(value) if (value := self.table_properties.get(config)) else default
```